### PR TITLE
WIP: Add England and Wales census tables

### PR DESCRIPTION
**Pull request by andrewphilipsmith**
This is WIP PR seeks to incorporate census attributes for England and Wales.

This supersedes #40 

Current Status:
------

* At present, each zipfile is downloaded multiple times. This is a consequence of the fact that each partition is a unique combination of topic summary and geometry level. Each zip file contains multiple files, including every geometry level for a given topic summary. I cannot see an easy way to share a cached downloaded zip file where different files are then extracted for different partitions (given dagster can in-theory execute partitions in arbitrary order and even on a cluster of machines). However a better solution might be to create a custom IOManager, but it is unclear if this is worth the effort at this stage.

Todo
-----
- [x] Implement `def _derived_metrics()` - This will need a very different implementation from the equivalent NI method. NI's source data is row-based, whereas EW is already column-based.

Problems to solve(d)
-----

- [x] Not all of the information required for the catalogue is available in on the bulk downloads webpage. The geometry level is only discoverable after downloading the zip file. We need to update the catalogue based on what we discover in the zip files.
    - [x] We build the partition on the (incorrect) assumption that all geometry levels are available for every source-table. We then allow those partitions for non-existing combinations of source-table and geometry-level to fail. 
- [x] The zip files can contain multiple CSV files, one for each geometry level. Do we only only what the lowest level geometry or include all of them?
    - [x] All of them - see above 
- [x] What is the appropriate way to handle the "extra geographies" column where data exists? Can these be easily combined with "original release" files?
    - The "extra geographies" are **_not_** included at present
    - [x] ~Decide whether or not to include the extra geometries in the `v0.2` release.~
- [x] Some of the downloaded files mistakenly have two consecutive `.` in the filename, e.g. `census2021-ts002-lsoa..csv`. We need to be able to gracefully handle this situation.
   - Done
- [x] There are multiple metrics in some source tables. We need to create the right (and right number of) MetricMetaData objects .
   - Stil to do; This will be handled by the `_derived_metrics()` method.
- [x] ~We need to find way to map the source metric names and HXL tags (possibly this should be in a separate PR)~

For future issues/PRs (eg after v0.2)
---

- [ ] Handle the "extra geographies" column where data exists
- [ ] Improve mapping of source metric names and HXL tags
- [ ] Root out repeated code with other countries

### CODE REVIEW

File: python/popgetter/assets/uk/england_wales_census.py [hunk 1]
```diff
@@ -0,0 +1,827 @@
...
+#    - the metadata object can be created for each table/metric
+from .united_kingdom import country
+
+
+@dataclass
+class EWCensusGeometryLevel:
+    level: str
+    geo_id_column: str
+    # TODO - I don't think this does anything meaningful
+    # census_table_column: str | None
```
Comment (sgreenbury) [inline python/popgetter/assets/uk/england_wales_census.py]: Yeah this was helpful for Northern Ireland but can be removed from this class if not needed.
```suggestion
```

File: python/popgetter/assets/uk/england_wales_census.py [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+            # in some of the filenames have two consecutive `.` in the filename
+            # e.g. `census2021-ts002-lsoa..csv`
+            extract_file_path = None
+            try:
+                extract_file_path = extract_main_file_from_zip(
+                    temp_zip, Path(temp_dir), source_archive_file_path
+                )
+            except ValueError:
+                source_archive_file_path = str(
+                    Path(source_archive_file_path).with_suffix("..csv")
```
Comment (sgreenbury) [inline python/popgetter/assets/uk/england_wales_census.py]:
```suggestion
                    Path(source_archive_file_path).with_suffix(".csv")
```

Comment (sgreenbury) [reply python/popgetter/assets/uk/england_wales_census.py]: Ah I see this is reporting the upstream problem!

Comment (andrewphilipsmith) [reply python/popgetter/assets/uk/england_wales_census.py]: The two dots are deliberate to deal with a bug I the source data. I've tried to provide a better explanation in the comments.

File: python/popgetter/assets/uk/england_wales_census.py (hunk 1 above)
Comment (andrewphilipsmith) [reply python/popgetter/assets/uk/england_wales_census.py]: I have removed this parameter throughout the code.

**Pull request merged**